flang/docs/tutorials/addingIntrinsics.md

```diff
@@ -0,0 +1,383 @@
+`lib/runtime/character.cpp` for an example.
+
+#### Tests
+To be defined. Ideally, we would want end-to-end tests from Fortran, but it is
+not yet clear where to put them in llvm. Unit tests can also be defined. See
+`unittests/Runtime/character.cpp` for an example.
+
+### Lowering
+- See: `lib/Lower/IntrinsicCall.cpp`
+- Where to submit patch: fir-dev branch of https://github.com/flang-compiler/f18-llvm-project.
+
+#### Implementation
+Lowering of intrinsic is driven by the `IntrinsicLibrary` class. Its member function
+`gen{intrinsic name}` lowers "intrinsic name". These functions have standardized interfaces.
+There is a table in this file that maps an intrinsic name to the function that lowers it.
+To add support for an intrinsic, a new member function containing the lowering implementation
+must be created and mapped into the table.
+
+The lowering implementation can be inlined if it can easily be implemented in a few fir or
+standard mlir dialect operations. Otherwise, the lowering implementation should generate code
+that emits a call to the runtime implementing the intrinsic.
+
+There are three standardized interfaces to choose from for implementation member functions:
+- `mlir::Value genXXX(mlir::Type, llvm::ArrayRef<mlir::Value>)`
+- `fir::ExtendedValue genXXX(mlir::Type, llvm::ArrayRef<fir::ExtendedValue>)`
+- `void genXXX(llvm::ArrayRef<fir::ExtendedValue>)`
+
+The first interface is to be used for elemental numerical and logical intrinsic functions,
+see `genAbs` for an example.
+The second interface is to be used for other intrinsic functions, see `genTrim` for an example.
+The third interface is to be used for intrinsic subroutine, see `genDateAndTime` for an example.
+
+It is worth being familiar with `fir::ExtendedValue`. They described an entity (symbol,
+or evaluated expression) with all its associated dynamic properties (length, extents, lower bounds...).
+A `fir::ExtendedValue` is a union of different SSA value containers that describe a particular category of entities.
+The categories are:
+- `fir::Unboxed` for scalar numericals, logicals, and non-polymorphic derived types without length parameters,
+- `fir::CharBoxValue` for scalar characters,
+- `fir::ArrayBoxValue` for contiguous arrays of numerical, logical, and non-polymorphic derived types without length parameters.
+- `fir::CharArrayBoxValue` contiguous character arrays,
+- `fir::MutableBoxValue` for allocatable and pointers of all types.
+- `fir::BoxValue` for all the rest.
+See `include/flang/Support/BoxValue.h` for more details.
+
+If the intrinsic implementation use runtime, the actual binding should be done in another helper function called
+`gen{Runtime function name}`. For instance character runtime bindings are implemented in `lib/Character/CharacterRuntime.cpp`
+See `genTrim` for an example of such runtime binding helper.
+Non character related runtime binding should be added to `lib/Character/Runtime.cpp`.
```

Comment:
Oops, these paths `/lib/Character/` were nonsense. Change to `/lib/Lower/`. Yes agree that Runtime.h should not be a miscellaneous bucket. Changed to `lib/Lower/IntrinsicRuntime.cpp` so that `Runtime.cpp` remains centered of execution related runtime (STOP and such).